knowhere/feder/HNSW.h

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#pragma once
+
+#include <memory>
+#include <tuple>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+
+#include "knowhere/feder/IDSet.h"
+
+namespace knowhere::feder::hnsw {
+
+///////////////////////////////////////////////////////////////////////////////
+// index view
+struct NodeInfo {
+    int64_t id_;
+    std::vector<int64_t> links_;
+
+    NodeInfo() = default;
+
+    NodeInfo(const int64_t id, const std::vector<int64_t>& links) = delete;
+
+    NodeInfo(const int64_t id, std::vector<int64_t>&& links) : id_(id), links_(links) {
+    }
+
+    NodeInfo(const NodeInfo& other) {
+        id_ = other.id_;
+        links_ = other.links_;
+    }
+
+    NodeInfo(NodeInfo&& other) {
+        id_ = other.id_;
+        links_ = std::move(other.links_);
+    }
+
+    NodeInfo
+    operator=(const NodeInfo& other) {
+        id_ = other.id_;
+        links_ = other.links_;
+        return *this;
+    }
+
+    NodeInfo
+    operator=(NodeInfo&& other) {
+        id_ = other.id_;
+        links_ = std::move(other.links_);
+        return *this;
+    }
+
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(NodeInfo, id_, links_)
+};
+
+struct LevelLinkGraph {
+    int64_t level_;
+    std::vector<NodeInfo> nodes_;
+
+    LevelLinkGraph() = default;
+
+    explicit LevelLinkGraph(int64_t level) : level_(level) {
+    }
+
+    LevelLinkGraph(const LevelLinkGraph& other) {
+        level_ = other.level_;
+        nodes_ = other.nodes_;
+    }
+
+    LevelLinkGraph(LevelLinkGraph&& other) {
+        level_ = other.level_;
+        nodes_ = std::move(other.nodes_);
+    }
+
+    LevelLinkGraph
+    operator=(const LevelLinkGraph& other) {
+        level_ = other.level_;
+        nodes_ = other.nodes_;
+        return *this;
+    }
+
+    LevelLinkGraph
+    operator=(LevelLinkGraph&& other) {
+        level_ = other.level_;
+        nodes_ = std::move(other.nodes_);
+        return *this;
+    }
+
+    void
+    AddNodeInfo(int64_t id, std::vector<int64_t>& links) {
+        nodes_.emplace_back(NodeInfo(id, std::move(links)));
+    }
+
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(LevelLinkGraph, level_, nodes_)
+};
+
+struct HNSWMeta {
```

Comment:
Same as before, class is preferred here. and we don't have to crowd them in a same file. Noticed that we have a very detail namespace like `feder::hnsw`, so no harm to create a folder under `feder` called `hnsw` and put all files there. This is just suggestion, depends on implementers preference.

Reply:
I prefer placing all HNSW related structures and classes in one file